src/src/com/microsoft/aad/adal/AuthenticationContext.java

```diff
@@ -1792,20 +1829,112 @@ private void setItemToCacheFromRefresh(final RefreshItem refreshItem,
             logReturnedToken(request, result);
 
             // Update for cache key
+            // If token is refreshed with Family token, make sure the tokenCacheItem doesn't
+            // contain the client id. 
+            final TokenCacheItem tokenCacheItem = new TokenCacheItem(request, result, refreshItem.mMultiResource);
+            if (tokenRefreshedWithFRT(request, refreshItem)) {
+                tokenCacheItem.setClientId(null);
+            } 
+            
             mTokenCacheStore.setItem(refreshItem.mKey, new TokenCacheItem(request, result,
                     refreshItem.mMultiResource));
-
             setItemToCache(request, result, false);
         }
     }
 
-    private void removeItemFromCache(final RefreshItem refreshItem) {
+    /**
+     * Remove token from cache. 
+     */
+    private void removeItemFromCache(final AuthenticationRequest authenticationRequest, final RefreshItem refreshItem) {
+        // Token removal logic:
+        // 1) If refresh with resource specific token cache entry, clear RT with key(R,C,U,A)
+        // 2) If refresh with MRRT, clear RT with (R,C,U,A) and (C,U,A)
+        // 3) if refresh with FRT, clear RT with (U,A)
+        if (tokenRefreshedWithRegularRT(authenticationRequest, refreshItem)) {
+            removeResourceSpecificTokenEntry(authenticationRequest, refreshItem);
+        } else if (tokenRefreshedWithMRRT(authenticationRequest, refreshItem)) {
+            removeMRRTokenEntry(authenticationRequest, refreshItem);
+            removeResourceSpecificTokenEntry(authenticationRequest, refreshItem);
+        } else if (tokenRefreshedWithFRT(authenticationRequest, refreshItem)) {
+            removeFamilyTokenEntry(authenticationRequest, refreshItem);
+        }
+    }
+    
+    /**
+     * @return True if access token redeem with RT is done through regular RT. 
+     */
+    private boolean tokenRefreshedWithRegularRT(final AuthenticationRequest authRequest, final RefreshItem refreshItem) {
```

Comment:
first, it doesn't make sense to create that many methods that used once or actually can be done in 1 line
1) check if refreshtoken is null only once in removeItemFromCache and return from function
2) if refreshItem had type of mKey it would be easier to compare in 1 line